src/linear_algebra/matrix/special_linear_group.lean

```diff
@@ -232,6 +232,31 @@ begin
   refl,
 end
 
+lemma fin_two_exists_eq_mk (g : SL(2, R)) :
```

Comment:
I think this would be more powerful if stated as an induction principle, something like
```
(h : ∀ a b c d (hdet : a * d - b * c = 1), P ⟨!![a, b; c, d], by rwa [matrix.det_fin_two_of]⟩) : ∀ x, P x
```

Reply:
Yes it seems so, I will try to make it more general.

Reply:
Good point, I guess the lemma would be:
```lean
lemma fin_two_induction (P : SL(2, R) → Prop)
  (h : ∀ a b c d (hdet : a * d - b * c = (1 : R)), P ⟨!![a, b; c, d], by rwa [det_fin_two_of]⟩)
  (g : SL(2, R)) : P g :=
begin
  obtain ⟨m, hm⟩ := g,
  convert h (m 0 0) (m 0 1) (m 1 0) (m 1 1) (by rwa det_fin_two at hm),
  ext i j, fin_cases i; fin_cases j; refl,
end
```

Reply:
It would be even easier if we had #15738

Reply:
`∀ (a b c d : R) (hdet : a * d - b * c = 1)` without the type ascription feels slightly clearer to me.

Reply:
I have added the lemma `fin_two_induction` but I have kept `fin_two_exists_eq_mk` based on the fact that I would otherwise need the following diff in the proof of `exists_SL2_smul_eq_of_apply_zero_one_ne_zero`:
```diff
-   obtain ⟨a, b, c, d, h, rfl⟩ := g.fin_two_exists_eq_mk,
+   revert h_denom hc,
+   refine fin_two_induction _ (λ a b c d h, _) g,
+   intros h_denom hc,
```
which persuades me that it's worth having both.

Reply:
Can't you do `induction g using the_namespace.fin_two_induction`?

Reply:
Ah yes.

I always forget about this feature of the `induction` tactic.